### PR TITLE
stubgen: Add support for yield statements

### PR DESCRIPTION
**Pull request by srittau**
### Description

Stubgen: Support functions that yield values:

```python
def foo():
    x = yield 123
    return ""
def bar():
    yield
```

will be generated as:

```python
from collections.abc import Generator
from typing import Any

def foo() -> Generator[Any, Any, Any]: ...
def bar() -> Generator[None, None, None]: ...
```

Before this change, these function were annotated with a `None` return type.

This PR doesn't support `yield from` expressions (see #10744).

## Test Plan

Tests for this feature and a few corner cases were added.

### CODE REVIEW

File: mypy/stubgen.py [hunk 1]
```diff
@@ -1107,6 +1131,14 @@ def add_typing_import(self, name: str) -> None:
         name = self.typing_name(name)
         self.import_tracker.require_name(name)
 
+    def add_abc_import(self, name: str) -> None:
+        """Add a name to be imported from collections.abc, unless it's imported already.
+
+        The import will be internal to the stub.
+        """
+        name = self.typing_name(name)
```
Comment (srittau) [inline mypy/stubgen.py:1139]: The name `typing_name` is not really correct anymore, since we use it for `collections.abc` as well. Should we rename it?

Comment (JelleZijlstra) [reply mypy/stubgen.py:1139]: I don't feel strongly, it's still a typing concept even if it lives in `collections.abc`.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by srittau** [comment]
Gentle pingedy pong.

**Comment by JelleZijlstra** [comment]
Thanks! I even ran into this when writing the beautifulsoup stubs the other day :)
